### ISSUE TITLE
Poor compilation of libc++ string comparison

### ISSUE THREAD

**Issue by bmaurer** [llvm/llvm-project#100898]
libc++'s std::string generates suboptimal machine code.

```
#include <string>

bool eqdynamic(const std::string& foo, const std::string& bar) {
    return foo == bar;
}

bool eqshort(const std::string& foo) { return foo == "aaa"; }

bool eqlong(const std::string& foo) {
    return foo == "asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf";
}
```

https://godbolt.org/z/xEWx59fns

* In the "dynamic" case, there is lots of branching to handle the combinations of short and long strings. The code might be simplified by testing if the inputs have the same "shortness" bit and then handling the differing case by normalizing the arguments so that the "short" input is the first input. if both strings are found to be short, it may be worth loading 8 bytes of each string and comparing the length and first 8 chars at the same time.
* In the "eqshort" case, once the string is found to be short, the first 4 bytes can be loaded and compared with [0x6, 'a', 'a', 'a']
* In the "eqlong" case we check if a "short" string can be 48 chars long, which is not possible, this branch could be eliminated statically.

```
eqdynamic(...):
        movzx   eax, byte ptr [rdi]
        test    al, 1
        je      .LBB0_1
        mov     rdx, qword ptr [rdi + 8]
        movzx   ecx, byte ptr [rsi]
        test    cl, 1
        jne     .LBB0_5
.LBB0_4:
        mov     r8d, ecx
        shr     r8d
        cmp     rdx, r8
        je      .LBB0_8
.LBB0_7:
        xor     eax, eax
        ret
.LBB0_1:
        mov     edx, eax
        shr     edx
        movzx   ecx, byte ptr [rsi]
        test    cl, 1
        je      .LBB0_4
.LBB0_5:
        mov     r8, qword ptr [rsi + 8]
        cmp     rdx, r8
        jne     .LBB0_7
.LBB0_8:
        test    al, 1
        je      .LBB0_9
        mov     rdi, qword ptr [rdi + 16]
        test    cl, 1
        jne     .LBB0_13
.LBB0_12:
        inc     rsi
        jmp     .LBB0_14
.LBB0_9:
        inc     rdi
        test    cl, 1
        je      .LBB0_12
.LBB0_13:
        mov     rsi, qword ptr [rsi + 16]
.LBB0_14:
        push    rax
        call    bcmp@PLT
        test    eax, eax
        sete    al
        add     rsp, 8
        ret

eqshort(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&):
        movzx   eax, byte ptr [rdi]
        test    al, 1
        je      .LBB1_1
        mov     rcx, qword ptr [rdi + 8]
        cmp     rcx, 3
        je      .LBB1_6
.LBB1_4:
        xor     eax, eax
        ret
.LBB1_1:
        mov     ecx, eax
        shr     ecx
        cmp     rcx, 3
        jne     .LBB1_4
.LBB1_6:
        test    al, 1
        je      .LBB1_7
        mov     rdi, qword ptr [rdi + 16]
        jmp     .LBB1_9
.LBB1_7:
        inc     rdi
.LBB1_9:
        movzx   eax, word ptr [rdi]
        xor     eax, 24929
        movzx   ecx, byte ptr [rdi + 2]
        xor     ecx, 97
        or      cx, ax
        sete    al
        ret

.LCPI2_0:
        .byte   97
...
.LCPI2_1:
        .byte   97   
...
eqlong(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&):
        movzx   eax, byte ptr [rdi]
        test    al, 1
        je      .LBB2_1
        mov     rcx, qword ptr [rdi + 8]
        cmp     rcx, 48
        je      .LBB2_6
.LBB2_4:
        xor     eax, eax
        ret
.LBB2_1:
        mov     ecx, eax
        shr     ecx
        cmp     rcx, 48
        jne     .LBB2_4
.LBB2_6:
        test    al, 1
        je      .LBB2_7
        mov     rdi, qword ptr [rdi + 16]
        jmp     .LBB2_9
.LBB2_7:
        inc     rdi
.LBB2_9:
        vmovdqu ymm0, ymmword ptr [rdi]
        vmovdqu xmm1, xmmword ptr [rdi + 32]
        vpxor   ymm0, ymm0, ymmword ptr [rip + .LCPI2_0]
        vpxor   ymm1, ymm1, ymmword ptr [rip + .LCPI2_1]
        vpor    ymm0, ymm0, ymm1
        vptest  ymm0, ymm0
        sete    al
        vzeroupper
        ret
```

**Comment by philnik777** [comment]
Could you elaborate on the dynamic case? I'm not sure what you're referring to. Regarding the short case, it looks to me like Clang missed to combine the compare and shift right, which results in the load not being combined. You might want to file an extra bug for this. Regarding the long case, that should be fairly trivial to fix. We can just add a `__builtin_constant_p` to check whether is a known constant and check that against the short string length. If it's larger than that, early return in case of a short string.

**Comment by bmaurer** [comment]
On the long case __builtin_assume(size < SMALL_SIZE) when getting the small size.

In the small case, it's also interesting that clang does `test    al, 1` at `LBB1_6`. At `LBB1_1` there's no reason to fall through to this check since we know it is equal. And then on the jump to LBB1_6 we know it is not equal.

On the dynamic case, maybe easier to look at a graph of the assembly 
<img width="646" alt="image" src="https://github.com/user-attachments/assets/821ffc79-f92e-4315-80ba-bb3ba862649b">

In the current version of the code there are 8 key basic blocks — 2 blocks for each of the 2 arguments for each of getting the length and getting the pointer. One way to minimize branching would be to simply get the length and pointer for each arg at the same time (even though the pointer is not needed if length is not equal)

But I also wonder if there's an approach where we leverage the fact that for small strings the length is part of the first byte. Something the first step is to xor the first byte of each string to compare the size (if the string is small) and the storage type all in one go. Might be a half baked idea

**Comment by philnik777** [comment]
I don't think the dynamic and short cases can be addressed in libc++. They seem to me like peephole optimizations to be fixed in the backend. For the long case I've created #100926.

**Comment by bmaurer** [comment]
This is a demo of an approach that substantially reduces branches:

https://godbolt.org/z/43cWYcsY3

As compared to the current approach:
* This packs the "is this string long" in the size field, that way one only needs to read from a single memory location to read the size
* Only two branches are required to detect if the strings are the same size, the first one should be extremely well predicted (a branch to detect if a string that could have been short is stored in a long format). For free, the equality of the first 7 chars of short strings are checked here.

**Comment by bmaurer** [comment]
(of course this is an ABI change which I assume libc++ is doesn't do very often. Maybe some of the same techniques are helpful in the current ABI. In particular xoring the "is long" bit can help reduce branch mispredicts if the common case is for the strings to be equal, as in a hash table)

**Comment by philnik777** [comment]
I consider this fixed with #100926, since the other cases are ABI breaking or should be handled by the compiler.
